Added docs for v2 transforms (part 1)

NicolasHug opened the pull request:
This PR adds docs for the v2 transforms that already exist in v1. I haven't tackled the new ones yet.

I jsut copy/pasted the docstrings and added:

- the `[BETA]` header in the first line
- the `.. betastatus::` sphinx directive
- replace "image" with "image / box / mask" where relevant.

NicolasHug commented on torchvision/transforms/v2/_geometry.py:
```diff
@@ -164,25 +305,23 @@ def _transform(self, inpt: Any, params: Dict[str, Any]) -> Any:
 class FiveCrop(Transform):
-    """
-    Example:
-        >>> class BatchMultiCrop(transforms.Transform):
```
The fact that `BatchMultiCrop`  was still there makes me think this example is largely outdated. I removed it, we can add it back if there's an up-to-date version somewhere.

pmeier replied:
Nope. I just updated it in the migration PR after we got rid of the labels datapoint. What makes you think it is outdated?

NicolasHug replied:
I went (way) too fast, sorry. I'll put it back.

pmeier commented on docs/source/transforms.rst:
```diff
@@ -162,9 +192,15 @@ Conversion
     :template: class.rst
 
     ToPILImage
+    v2.ToPILImage
```
We only kept this for BC but prefer `ToImagePIL` since it aligns better with our naming scheme. Should we keep it out of the documentation?

pmeier replied:
> From a user perspective, I don't think "I should be updating my code to remove ToPILImage and use ToImagePIL instead because the torchvision maintainers have a new naming scheme" is a super compelling story.

Agreed. This is why we left it in as alias:

https://github.com/pytorch/vision/blob/928b05cad36eadb13e169f03028767c8bcd1f21d/torchvision/transforms/v2/_type_conversion.py#L72-L74

Meaning, only users who don't have old pipelines are going to see this and there it feels ok to not show the old transform.

In case you still want to, can we maybe just let the docstring say "alias of X" like PyTorch core is doing in such situations? For example:

https://pytorch.org/docs/stable/generated/torch.not_equal.html?highlight=not_equal#torch.not_equal

NicolasHug replied:
yeah, this is a sphinx built-in behaviour I believe, and this is what we already have:

![image](https://user-images.githubusercontent.com/1190450/220408025-9bfa818f-254e-43ca-8f89-477bbd4b4de9.png)


![image](https://user-images.githubusercontent.com/1190450/220408060-d7b72e2b-8910-4af3-ad1b-e11f77fdc1f5.png)

pmeier commented on docs/source/transforms.rst:
```diff
@@ -162,9 +192,15 @@ Conversion
+    v2.ToPILImage
+    v2.ToImagePIL
     ToTensor
+    v2.ToTensor
```
Since it is deprecated, should we document it?

pmeier replied:
It seems that `ToImageTensor` is missing from the list.

NicolasHug replied:
> Since it is deprecated, should we document it?

Some thoughts here https://github.com/pytorch/vision/pull/7297/files#r1113244517

> It seems that ToImageTensor is missing from the list.

I did not add any new v2 transforms here, so that's normal

pmeier commented on torchvision/transforms/v2/_augment.py:
```diff
@@ -12,7 +12,38 @@
 from .utils import is_simple_tensor, query_chw
 
 
+# TODO: Just move that to _misc.py?
```
Previously, `CutMix` and `MixUp` where also here. So we'll have to move all three. IMO `_augment` is fine albeit a little generic.

pmeier commented on torchvision/transforms/v2/_color.py:
```diff
@@ -11,6 +11,23 @@
 
 
 class Grayscale(Transform):
+    """[BETA] Convert image to grayscale.
```
Applies to every color transform

NicolasHug replied:
(un-resolving because I didn't apply that to the rest of the color transforms yet)

pmeier commented on torchvision/transforms/v2/_container.py:
```diff
@@ -9,6 +9,37 @@
+
+    .. betastatus:: Compose transform
+
+    This transform does not support torchscript.
```
Should we advertise this at all given that we want to get rid of it?

NicolasHug replied:
I think so... But that makes me think we also want to document the JIT thingy in most transforms as well. I suggest we add a "JIT support for v2 transforms" in our docs somewhere, and then automatically link to it in the (edited) warning message of the `.. betastatus::` directive?

pmeier replied:
Fine by me.

pmeier commented on torchvision/transforms/v2/_container.py:
```diff
@@ -29,6 +60,27 @@ def extra_repr(self) -> str:
+    .. betastatus:: RandomApply transform
+
+    .. note::
+        In order to script the transformation, please use ``torch.nn.ModuleList`` as input instead of list/tuple of
```
Same as above

pmeier commented on torchvision/transforms/v2/_deprecated.py:
```diff
@@ -10,6 +10,31 @@
+    .. betastatus:: ToTensor transform
+
+    .. warning::
+        v2.ToTensor is deprecated and will be removed in a future release.
```
```suggestion
        :class:`v2.ToTensor` is deprecated and will be removed in a future release.
```

pmeier commented on torchvision/transforms/v2/_geometry.py:
```diff
@@ -26,20 +26,100 @@
 
 
 class RandomHorizontalFlip(_RandomApplyTransform):
+    """[BETA] Horizontally flip the given image/box/mask randomly with a given probability.
```
video is missing here as well

pmeier commented on torchvision/transforms/v2/_geometry.py:
```diff
@@ -209,8 +348,25 @@ def _check_inputs(self, flat_inputs: List[Any]) -> None:
 
 class TenCrop(Transform):
-    """
-    See :class:`~torchvision.transforms.v2.FiveCrop` for an example.
```
If we reinstate the example above, we should also reinstate this.

pmeier commented on torchvision/transforms/v2/_geometry.py:
```diff
@@ -443,6 +703,52 @@ def _transform(self, inpt: Any, params: Dict[str, Any]) -> Any:
+            on left/right and top/bottom respectively. If a sequence of length 4 is provided
+            this is the padding for the left, top, right and bottom borders respectively.
+
+            .. note::
+                In torchscript mode padding as single int is not supported, use a sequence of
+                length 1: ``[padding, ]``.
```
This comes up multiple times. Same question as above: should we document this at all?

pmeier commented on torchvision/transforms/v2/_misc.py:
```diff
@@ -124,7 +174,26 @@ def _transform(
         return F.normalize(inpt, mean=self.mean, std=self.std, inplace=self.inplace)
 
 
+# TODO: This should be in the _color.py file!
```
I don't feel it belongs there, but no strong opinion.

NicolasHug replied:
But the transform is in under the "color" section in the docs?

pmeier replied:
Oh, in that case I missed to flag this there.

vfdev-5 commented on torchvision/transforms/v2/_augment.py:
```diff
@@ -12,7 +12,38 @@
 
+# TODO: Just move that to _misc.py?
 class RandomErasing(_RandomApplyTransform):
+    """[BETA] Randomly selects a rectangle region in a torch.Tensor image and erases its pixels.
```
```suggestion
    """[BETA] Randomly selects a rectangle region in the input image or video and erases its pixels.
```

vfdev-5 reviewed (approved):
stamped

NicolasHug commented:
Thanks @pmeier @vfdev-5 for the review - I'll merge now to unblock upcoming PRs, but there are outstanding comments that I haven't addressed (yet) in this PR. We'll address them in the next few PRs to come. I marked the alread-addressed comments as resolved, so we just need to check the un-resolved ones.
